packages/block-library/src/navigation/edit.js

```diff
@@ -126,56 +182,181 @@ function Navigation( {
 	}
 
 	function handleCreateFromExistingPages() {
-		updateNavItemBlocks( defaultPagesNavigationItems );
+		updateNavItemBlocks( navLinkBlocksFromPages );
 		selectBlock( clientId );
 	}
 
-	const hasPages = hasResolvedPages && pages && pages.length;
+	function handleCreateFromExistingMenu() {
+		updateNavItemBlocks( navLinkBlocksFromMenuItems );
+		selectBlock( clientId );
+	}
 
-	const blockInlineStyles = {
-		fontSize: fontSize.size ? fontSize.size + 'px' : undefined,
-	};
+	function handleCreate() {
+		const { key } = selectedDropDownOption;
+
+		// Explicity request to create empty.
+		if ( key === CREATE_EMPTY_OPTION_VALUE ) {
+			return handleCreateEmpty();
+		}
+
+		// Create from Pages.
+		if ( hasPages && key === CREATE_FROM_PAGES_OPTION_VALUE ) {
+			return handleCreateFromExistingPages();
+		}
+
+		// Create from WP Menu (if exists and not empty).
+		if (
+			hasMenus &&
+			selectedDropDownOption &&
+			navLinkBlocksFromMenuItems?.length
+		) {
+			return handleCreateFromExistingMenu();
+		}
+
+		// Default to empty menu
+		return handleCreateEmpty();
+	}
+
+	function buildPlaceHolderInstructionText() {
+		if ( hasMenus && hasPages ) {
+			return __(
+				'Create a Navigation from all existing pages, or chose a Menu.'
+			);
+		}
+
+		if ( ! hasMenus && ! hasPages ) {
+			return __( 'Create a an empty Navigation.' );
+		}
+
+		if ( hasMenus && ! hasPages ) {
+			return __( 'Create a Navigation from a Menu or create empty.' );
+		}
+
+		if ( ! hasMenus && hasPages ) {
+			return __(
+				'Create a Navigation from all existing pages, or create empty.'
+			);
+		}
+	}
+
+	const dropDownOptions = [
+		{
+			id: CREATE_PLACEHOLDER_VALUE,
+			name: __( 'Select where to start from…' ),
+		},
+		...( hasMenus ? menus : [] ),
+		{
+			id: CREATE_EMPTY_OPTION_VALUE,
+			name: __( 'Create empty menu' ),
+		},
+		...( hasPages
+			? [
+					{
+						id: CREATE_FROM_PAGES_OPTION_VALUE,
+						name: __( 'New from all top-level pages' ),
+					},
+			  ]
+			: [] ),
+	];
+
+	function shouldDisableCreateButton() {
```

Comment:
I think these functions should be extracted, made pure, and then memoized using `useCallback`.

This PR makes the Navigation edit function quite big.

Reply:
✅ Memoized.

I'm trying to wrap my head around how you envisage "extracting" these. Are you imagining them in separate files? Outside the Navigation component? They would then need to be pure by design right?